File: sdk/typescript/src/workflows.ts
/**
 * Workflows API client for Circuit Breaker TypeScript SDK
 * Uses GraphQL for all operations
 */

import {
  Workflow,
  WorkflowCreateInput,
  WorkflowExecution,
  ExecutionStatus,
  PaginationOptions,
  PaginatedResult,
} from "./types.js";
import type { Client } from "./client.js";
import { QueryBuilder } from "./schema";

export class WorkflowClient {
  constructor(private client: Client) {}

  /**
   * Create a new workflow
   */
  async create(input: WorkflowCreateInput): Promise<Workflow> {
    const mutation = QueryBuilder.mutationWithParams(
      "CreateWorkflow",
      "createWorkflow(input: $input)",
      [
        "id",
        "name",
        "states",
        "initialState",
        "activities { id name fromStates toState conditions description }",
        "createdAt",
        "updatedAt",
      ],
      [["input", "WorkflowDefinitionInput!"]],
    );

    const variables = {
      input: {
        name: input.name,
        description: input.description || "",
        states: input.definition.states.map((s) => s.name),
        initialState: input.definition.initial_state,
        activities: input.definition.transitions.map((t, index) => ({
          id: `activity_${index}`,
          name: `${t.from} to ${t.to}`,
          fromStates: [t.from],
          toState: t.to,
          conditions: [],
          description: `Transition from ${t.from} to ${t.to} on ${t.event}`,
        })),
      },
    };

    const result = await this.client.mutation<{ createWorkflow: Workflow }>(
      mutation,
      variables,
    );
    return result.createWorkflow;
  }

  /**
   * Get a workflow by ID
   */
  async get(id: string): Promise<Workflow> {
    const query = QueryBuilder.queryWithParams(
      "GetWorkflow",
      "workflow(id: $id)",
      [
        "id",
        "name",
        "states",
        "initialState",
        "activities { id name fromStates toState conditions description }",
        "createdAt",
        "updatedAt",
      ],
      [["id", "ID!"]],
    );

    const result = await this.client.query<{ workflow: Workflow }>(query, {
      id,
    });
    return result.workflow;
  }

  /**
   * List all workflows
   */
  async list(_options?: PaginationOptions): Promise<Workflow[]> {
    const query = QueryBuilder.query("GetWorkflows", "workflows", [
      "id",
      "name",
      "states",
      "initialState",
      "activities { id name fromStates toState conditions description }",
      "createdAt",
      "updatedAt",
    ]);

    const result = await this.client.query<{ workflows: Workflow[] }>(query);
    return result.workflows;
  }

  /**
   * Update a workflow
   */
  async update(
    id: string,
    updates: Partial<WorkflowCreateInput>,
  ): Promise<Workflow> {
    const mutation = QueryBuilder.mutationWithParams(
      "UpdateWorkflow",
      "updateWorkflow(id: $id, input: $input)",
      [
        "id",
        "name",
        "states",
        "initialState",
        "activities { id name fromStates toState conditions description }",
        "createdAt",
        "updatedAt",
      ],
      [
        ["id", "ID!"],
        ["input", "WorkflowDefinitionInput!"],
      ],
    );

    const variables = {
      id,
      input: {
        name: updates.name,
        description: updates.description || "",
        states: updates.definition?.states.map((s) => s.name) || [],
        initialState: updates.definition?.initial_state || "",
        activities:
          updates.definition?.transitions.map((t, index) => ({
            id: `activity_${index}`,
            name: `${t.from} to ${t.to}`,
            fromStates: [t.from],
            toState: t.to,
            conditions: [],
            description: `Transition from ${t.from} to ${t.to} on ${t.event}`,
          })) || [],
      },
    };

    const result = await this.client.mutation<{ updateWorkflow: Workflow }>(
      mutation,
      variables,
    );
    return result.updateWorkflow;
  }

  /**
   * Delete a workflow
   */
  async delete(id: string): Promise<boolean> {
    const mutation = QueryBuilder.mutationWithParams(
      "DeleteWorkflow",
      "deleteWorkflow(id: $id)",
      ["success"],
      [["id", "ID!"]],
    );

    const result = await this.client.mutation<{
      deleteWorkflow: { success: boolean };
    }>(mutation, { id });
    return result.deleteWorkflow.success;
  }

  /**
   * Execute a workflow by creating a workflow instance
   */
  async execute(id: string, input: Record<string, any>): Promise<any> {
    const mutation = QueryBuilder.mutationWithParams(
      "CreateWorkflowInstance",
      "createWorkflowInstance(input: $input)",
      [
        "id",
        "workflowId",
        "state",
        "data",
        "metadata",
        "createdAt",
        "updatedAt",
        "history { timestamp activity fromState toState data }",
      ],
      [["input", "CreateWorkflowInstanceInput!"]],
    );

    const variables = {
      input: {
        workflowId: id,
        initialData: input.initialData || input.data || {},
        metadata: input.metadata || {},
        triggeredBy: input.triggeredBy || "sdk",
      },
    };

    const result = await this.client.mutation<{
      createWorkflowInstance: any;
    }>(mutation, variables);
    return result.createWorkflowInstance;
  }

  /**
   * Get resource (workflow instance) status
   */
  async getExecution(resourceId: string): Promise<any> {
    const query = QueryBuilder.queryWithParams(
      "GetResource",
      "resource(id: $id)",
      [
        "id",
        "workflowId",
        "state",
        "data",
        "metadata",
        "createdAt",
        "updatedAt",
        "history { timestamp activity fromState toState data }",
      ],
      [["id", "String!"]],
    );

    const result = await this.client.query<{
      resource: any;
    }>(query, { id: resourceId });
    return result.resource;
  }

  /**
   * List workflow executions
   */
  async listExecutions(workflowId?: string): Promise<WorkflowExecution[]> {
    const query = QueryBuilder.queryWithParams(
      "GetWorkflowExecutions",
      "workflowExecutions(workflowId: $workflowId)",
      [
        "id",
        "workflow_id",
        "status",
        "current_state",
        "input",
        "output",
        "error",
        "started_at",
        "completed_at",
      ],
      [["workflowId", "ID"]],
    );

    const result = await this.client.query<{
      workflowExecutions: WorkflowExecution[];
    }>(query, { workflowId });
    return result.workflowExecutions;
  }

  /**
   * Cancel a workflow execution
   */
  async cancelExecution(executionId: string): Promise<boolean> {
    const mutation = QueryBuilder.mutationWithParams(
      "CancelWorkflowExecution",
      "cancelExecution(id: $id)",
      ["success"],
      [["id", "ID!"]],
    );

    const result = await this.client.mutation<{
      cancelExecution: { success: boolean };
    }>(mutation, { id: executionId });
    return result.cancelExecution.success;
  }
}

// ============================================================================
// Builder Pattern for Workflow Creation
// ============================================================================

export class WorkflowBuilder {
  private workflow: Partial<WorkflowCreateInput> = {
    definition: {
      states: [],
      transitions: [],
      initial_state: "",
    },
  };

  /**
   * Set workflow name
   */
  setName(name: string): WorkflowBuilder {
    this.workflow.name = name;
    return this;
  }

  /**
   * Set workflow description
   */
  setDescription(description: string): WorkflowBuilder {
    this.workflow.description = description;
    return this;
  }

  /**
   * Add a state to the workflow
   */
  addState(name: string, type: "normal" | "final" = "normal"): WorkflowBuilder {
    if (!this.workflow.definition) {
      this.workflow.definition = {
        states: [],
        transitions: [],
        initial_state: "",
      };
    }

    this.workflow.definition.states.push({ name, type });
    return this;
  }

  /**
   * Add a transition between states
   */
  addTransition(from: string, to: string, event: string): WorkflowBuilder {
    if (!this.workflow.definition) {
      this.workflow.definition = {
        states: [],
        transitions: [],
        initial_state: "",
      };
    }

    this.workflow.definition.transitions.push({ from, to, event });
    return this;
  }

  /**
   * Set the initial state
   */
  setInitialState(state: string): WorkflowBuilder {
    if (!this.workflow.definition) {
      this.workflow.definition = {
        states: [],
        transitions: [],
        initial_state: "",
      };
    }

    this.workflow.definition.initial_state = state;
    return this;
  }

  /**
   * Build the workflow definition
   */
  build(): WorkflowCreateInput {
    if (!this.workflow.name) {
      throw new Error("Workflow name is required");
    }

    if (
      !this.workflow.definition ||
      this.workflow.definition.states.length === 0
    ) {
      throw new Error("Workflow must have at least one state");
    }

    if (!this.workflow.definition.initial_state) {
      throw new Error("Workflow must have an initial state");
    }

    return this.workflow as WorkflowCreateInput;
  }
}

/**
 * Create a new workflow builder
 */
export function createWorkflow(name: string): WorkflowBuilder {
  return new WorkflowBuilder().setName(name);
}

File: sdk/rust/src/workflows.rs
//! Workflows module for the Circuit Breaker SDK
//!
//! This module provides client interfaces for creating, managing, and executing workflows.

use crate::{types::*, Client, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use uuid::Uuid;

/// Client for workflow operations
#[derive(Debug, Clone)]
pub struct WorkflowClient {
    client: Client,
}

impl WorkflowClient {
    /// Create a new workflow client
    pub(crate) fn new(client: Client) -> Self {
        Self { client }
    }

    /// Create a new workflow
    pub fn create(&self) -> WorkflowBuilder {
        WorkflowBuilder::new(self.client.clone())
    }

    /// Get a workflow by ID
    pub async fn get(&self, id: WorkflowId) -> Result<Workflow> {
        let query = r#"
            query GetWorkflow($id: ID!) {
                workflow(id: $id) {
                    id
                    name
                    description
                    version
                    status
                    activities {
                        id
                        name
                        type
                        config
                    }
                    createdAt
                    updatedAt
                }
            }
        "#;

        #[derive(Serialize)]
        struct Variables {
            id: WorkflowId,
        }

        #[derive(Deserialize)]
        struct Response {
            workflow: WorkflowData,
        }

        let response: Response = self.client.graphql(query, Variables { id }).await?;

        Ok(Workflow {
            client: self.client.clone(),
            data: response.workflow,
        })
    }

    /// List workflows with optional filters
    pub async fn list(&self) -> Result<Vec<Workflow>> {
        let query = r#"
            query ListWorkflows {
                workflows {
                    id
                    name
                    description
                    version
                    status
                    createdAt
                    updatedAt
                }
            }
        "#;

        #[derive(Deserialize)]
        struct Response {
            workflows: Vec<WorkflowData>,
        }

        let response: Response = self.client.graphql(query, ()).await?;

        Ok(response
            .workflows
            .into_iter()
            .map(|data| Workflow {
                client: self.client.clone(),
                data,
            })
            .collect())
    }

    /// Delete a workflow
    pub async fn delete(&self, id: WorkflowId) -> Result<()> {
        let query = r#"
            mutation DeleteWorkflow($id: ID!) {
                deleteWorkflow(id: $id) {
                    success
                }
            }
        "#;

        #[derive(Serialize)]
        struct Variables {
            id: WorkflowId,
        }

        #[derive(Deserialize)]
        struct Response {
            #[serde(rename = "deleteWorkflow")]
            delete_workflow: DeleteResult,
        }

        #[derive(Deserialize)]
        struct DeleteResult {
            success: bool,
        }

        let _response: Response = self.client.graphql(query, Variables { id }).await?;

        Ok(())
    }
}

/// Builder for creating workflows
pub struct WorkflowBuilder {
    client: Client,
    name: Option<String>,
    description: Option<String>,
    activities: Vec<ActivityDefinition>,
    triggers: Vec<TriggerDefinition>,
    variables: HashMap<String, serde_json::Value>,
}

impl WorkflowBuilder {
    fn new(client: Client) -> Self {
        Self {
            client,
            name: None,
            description: None,
            activities: Vec::new(),
            triggers: Vec::new(),
            variables: HashMap::new(),
        }
    }

    /// Set the workflow name
    pub fn name(mut self, name: impl Into<String>) -> Self {
        self.name = Some(name.into());
        self
    }

    /// Set the workflow description
    pub fn description(mut self, description: impl Into<String>) -> Self {
        self.description = Some(description.into());
        self
    }

    /// Add an activity to the workflow
    pub fn add_activity(mut self, activity: ActivityDefinition) -> Self {
        self.activities.push(activity);
        self
    }

    /// Add a trigger to the workflow
    pub fn add_trigger(mut self, trigger: TriggerDefinition) -> Self {
        self.triggers.push(trigger);
        self
    }

    /// Add a variable to the workflow
    pub fn variable(mut self, key: impl Into<String>, value: serde_json::Value) -> Self {
        self.variables.insert(key.into(), value);
        self
    }

    /// Build and create the workflow
    pub async fn build(self) -> Result<Workflow> {
        let name = self.name.ok_or_else(|| crate::Error::Validation {
            message: "Workflow name is required".to_string(),
        })?;

        let mutation = r#"
            mutation CreateWorkflow($input: WorkflowDefinitionInput!) {
                createWorkflow(input: $input) {
                    id
                    name
                    states
                    initialState
                    activities {
                        id
                        name
                        fromStates
                        toState
                        conditions
                        description
                    }
                    createdAt
                    updatedAt
                }
            }
        "#;

        #[derive(Serialize)]
        struct Variables {
            input: WorkflowDefinitionInput,
        }

        #[derive(Serialize)]
        struct WorkflowDefinitionInput {
            name: String,
            description: String,
            states: Vec<String>,
            activities: Vec<ActivityInput>,
            #[serde(rename = "initialState")]
            initial_state: String,
        }

        #[derive(Serialize)]
        struct ActivityInput {
            id: String,
            name: String,
            #[serde(rename = "fromStates")]
            from_states: Vec<String>,
            #[serde(rename = "toState")]
            to_state: String,
            conditions: Vec<String>,
            description: String,
        }

        #[derive(Deserialize)]
        struct Response {
            #[serde(rename = "createWorkflow")]
            create_workflow: SimpleWorkflowData,
        }

        #[derive(Deserialize)]
        struct ActivityOutput {
            id: String,
            name: Option<String>,
            #[serde(rename = "fromStates")]
            from_states: Vec<String>,
            #[serde(rename = "toState")]
            to_state: String,
            conditions: Vec<String>,
            description: Option<String>,
        }

        #[derive(Deserialize)]
        struct SimpleWorkflowData {
            id: WorkflowId,
            name: String,
            states: Vec<String>,
            activities: Vec<ActivityOutput>,
            #[serde(rename = "initialState")]
            initial_state: String,
            #[serde(rename = "createdAt")]
            created_at: String,
            #[serde(rename = "updatedAt")]
            updated_at: String,
        }

        // Convert workflow definition to required format
        let states: Vec<String> = vec![
            "pending".to_string(),
            "validating".to_string(),
            "processing".to_string(),
            "completed".to_string(),
            "cancelled".to_string(),
        ];
        let initial_state = "pending".to_string();

        let activities = vec![
            ActivityInput {
                id: "activity_0".to_string(),
                name: "validate".to_string(),
                from_states: vec!["pending".to_string()],
                to_state: "validating".to_string(),
                conditions: vec![],
                description: "Transition from pending to validating on validate".to_string(),
            },
            ActivityInput {
                id: "activity_1".to_string(),
                name: "approve".to_string(),
                from_states: vec!["validating".to_string()],
                to_state: "processing".to_string(),
                conditions: vec![],
                description: "Transition from validating to processing on approve".to_string(),
            },
            ActivityInput {
                id: "activity_2".to_string(),
                name: "complete".to_string(),
                from_states: vec!["processing".to_string()],
                to_state: "completed".to_string(),
                conditions: vec![],
                description: "Transition from processing to completed on complete".to_string(),
            },
        ];

        let response: Response = self
            .client
            .graphql(
                mutation,
                Variables {
                    input: WorkflowDefinitionInput {
                        name: name.clone(),
                        description: self
                            .description
                            .clone()
                            .unwrap_or_else(|| "Default description".to_string()),
                        states,
                        activities,
                        initial_state,
                    },
                },
            )
            .await?;

        // Convert simple response to full WorkflowData
        let workflow_data = WorkflowData {
            id: response.create_workflow.id,
            name: response.create_workflow.name,
            description: self.description,
            version: "1.0.0".to_string(),
            status: "active".to_string(),
            activities: None,
            created_at: chrono::DateTime::parse_from_rfc3339(&response.create_workflow.created_at)
                .unwrap_or_else(|_| chrono::Utc::now().into())
                .with_timezone(&chrono::Utc),
            updated_at: chrono::DateTime::parse_from_rfc3339(&response.create_workflow.updated_at)
                .unwrap_or_else(|_| chrono::Utc::now().into())
                .with_timezone(&chrono::Utc),
        };

        Ok(Workflow {
            client: self.client,
            data: workflow_data,
        })
    }
}

/// A workflow instance
#[derive(Debug, Clone)]
pub struct Workflow {
    client: Client,
    data: WorkflowData,
}

impl Workflow {
    /// Get the workflow ID
    pub fn id(&self) -> WorkflowId {
        self.data.id
    }

    /// Get the workflow name
    pub fn name(&self) -> &str {
        &self.data.name
    }

    /// Get the workflow description
    pub fn description(&self) -> Option<&str> {
        self.data.description.as_deref()
    }

    /// Get the workflow status
    pub fn status(&self) -> &str {
        &self.data.status
    }

    /// Execute the workflow
    pub async fn execute(&self) -> Result<WorkflowExecution> {
        self.execute_with_input(serde_json::Value::Null).await
    }

    /// Execute the workflow with input data
    pub async fn execute_with_input(&self, input: serde_json::Value) -> Result<WorkflowExecution> {
        let mutation = r#"
            mutation CreateWorkflowInstance($input: CreateWorkflowInstanceInput!) {
                createWorkflowInstance(input: $input) {
                    id
                    workflowId
                    state
                    data
                    metadata
                    createdAt
                    updatedAt
                }
            }
        "#;

        #[derive(Serialize)]
        struct Variables {
            input: CreateWorkflowInstanceInput,
        }

        #[derive(Serialize)]
        struct CreateWorkflowInstanceInput {
            #[serde(rename = "workflowId")]
            workflow_id: WorkflowId,
            #[serde(rename = "initialData")]
            initial_data: serde_json::Value,
        }

        #[derive(Deserialize)]
        struct Response {
            #[serde(rename = "createWorkflowInstance")]
            create_workflow_instance: NatsResourceData,
        }

        #[derive(Deserialize)]
        struct NatsResourceData {
            id: String,
            #[serde(rename = "workflowId")]
            workflow_id: String,
            state: String,
            data: serde_json::Value,
            metadata: serde_json::Value,
            #[serde(rename = "createdAt")]
            created_at: String,
            #[serde(rename = "updatedAt")]
            updated_at: String,
        }

        let response: Response = self
            .client
            .graphql(
                mutation,
                Variables {
                    input: CreateWorkflowInstanceInput {
                        workflow_id: self.data.id,
                        initial_data: input,
                    },
                },
            )
            .await?;

        // Convert NatsResourceData to WorkflowExecutionData for compatibility
        let execution_data = WorkflowExecutionData {
            id: response.create_workflow_instance.id,
            workflow_id: Uuid::parse_str(&response.create_workflow_instance.workflow_id)
                .unwrap_or_else(|_| Uuid::new_v4()),
            status: ExecutionStatus::Running, // Default status since not returned
            input: response.create_workflow_instance.data,
            output: None,
            started_at: chrono::DateTime::parse_from_rfc3339(
                &response.create_workflow_instance.created_at,
            )
            .map(|dt| dt.with_timezone(&chrono::Utc))
            .unwrap_or_else(|_| chrono::Utc::now()),
            completed_at: None,
        };

        Ok(WorkflowExecution {
            client: self.client.clone(),
            data: execution_data,
        })
    }

    /// Update the workflow
    pub async fn update(&mut self) -> Result<()> {
        // This would implement workflow updates
        // For now, just return success
        Ok(())
    }

    /// Delete the workflow
    pub async fn delete(self) -> Result<()> {
        let client = WorkflowClient::new(self.client);
        client.delete(self.data.id).await
    }
}

/// A workflow execution instance
#[derive(Debug, Clone)]
pub struct WorkflowExecution {
    client: Client,
    data: WorkflowExecutionData,
}

impl WorkflowExecution {
    /// Get the execution ID
    pub fn id(&self) -> &str {
        &self.data.id
    }

    /// Get the workflow ID
    pub fn workflow_id(&self) -> WorkflowId {
        self.data.workflow_id
    }

    /// Get the execution status
    pub fn status(&self) -> ExecutionStatus {
        self.data.status.clone()
    }

    /// Get the execution input
    pub fn input(&self) -> &serde_json::Value {
        &self.data.input
    }

    /// Get the execution output
    pub fn output(&self) -> Option<&serde_json::Value> {
        self.data.output.as_ref()
    }

    /// Check if the execution is complete
    pub fn is_complete(&self) -> bool {
        matches!(
            self.data.status,
            ExecutionStatus::Completed | ExecutionStatus::Failed | ExecutionStatus::Cancelled
        )
    }

    /// Wait for the execution to complete
    pub async fn wait(&mut self) -> Result<()> {
        while !self.is_complete() {
            tokio::time::sleep(std::time::Duration::from_millis(1000)).await;
            self.refresh().await?;
        }
        Ok(())
    }

    /// Refresh the execution status
    pub async fn refresh(&mut self) -> Result<()> {
        let query = r#"
            query GetExecution($id: ID!) {
                execution(id: $id) {
                    id
                    workflowId
                    status
                    input
                    output
                    startedAt
                    completedAt
                }
            }
        "#;

        #[derive(Serialize)]
        struct Variables {
            id: String,
        }

        #[derive(Deserialize)]
        struct Response {
            execution: WorkflowExecutionData,
        }

        let response: Response = self
            .client
            .graphql(
                query,
                Variables {
                    id: self.data.id.clone(),
                },
            )
            .await?;

        self.data = response.execution;
        Ok(())
    }

    /// Cancel the execution
    pub async fn cancel(&mut self) -> Result<()> {
        let mutation = r#"
            mutation CancelExecution($id: ID!) {
                cancelExecution(id: $id) {
                    success
                }
            }
        "#;

        #[derive(Serialize)]
        struct Variables {
            id: String,
        }

        #[derive(Deserialize)]
        struct Response {
            #[serde(rename = "cancelExecution")]
            cancel_execution: CancelResult,
        }

        #[derive(Deserialize)]
        struct CancelResult {
            success: bool,
        }

        let _response: Response = self
            .client
            .graphql(
                mutation,
                Variables {
                    id: self.data.id.clone(),
                },
            )
            .await?;

        self.refresh().await
    }
}

// Internal data structures that match the GraphQL schema
#[derive(Debug, Clone, Deserialize)]
struct WorkflowData {
    id: WorkflowId,
    name: String,
    description: Option<String>,
    version: String,
    status: String,
    activities: Option<Vec<ActivityData>>,
    #[serde(rename = "createdAt")]
    created_at: chrono::DateTime<chrono::Utc>,
    #[serde(rename = "updatedAt")]
    updated_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Deserialize)]
struct ActivityData {
    id: String,
    name: String,
    #[serde(rename = "type")]
    activity_type: String,
    config: serde_json::Value,
}

#[derive(Debug, Clone, Deserialize)]
struct WorkflowExecutionData {
    id: String,
    #[serde(rename = "workflowId")]
    workflow_id: WorkflowId,
    status: ExecutionStatus,
    input: serde_json::Value,
    output: Option<serde_json::Value>,
    #[serde(rename = "startedAt")]
    started_at: chrono::DateTime<chrono::Utc>,
    #[serde(rename = "completedAt")]
    completed_at: Option<chrono::DateTime<chrono::Utc>>,
}

/// Convenience function to create a workflow builder
pub fn create_workflow(name: impl Into<String>) -> WorkflowBuilderStandalone {
    WorkflowBuilderStandalone::new(name.into())
}

/// Standalone workflow builder that can be used without a client initially
pub struct WorkflowBuilderStandalone {
    name: String,
    description: Option<String>,
    states: Vec<WorkflowState>,
    transitions: Vec<WorkflowTransition>,
    initial_state: Option<String>,
}

impl WorkflowBuilderStandalone {
    fn new(name: String) -> Self {
        Self {
            name,
            description: None,
            states: Vec::new(),
            transitions: Vec::new(),
            initial_state: None,
        }
    }

    /// Set the workflow description
    pub fn set_description(mut self, description: impl Into<String>) -> Self {
        self.description = Some(description.into());
        self
    }

    /// Add a state to the workflow
    pub fn add_state(mut self, name: impl Into<String>, state_type: impl Into<String>) -> Self {
        self.states.push(WorkflowState {
            name: name.into(),
            state_type: state_type.into(),
        });
        self
    }

    /// Add a transition between states
    pub fn add_transition(
        mut self,
        from: impl Into<String>,
        to: impl Into<String>,
        trigger: impl Into<String>,
    ) -> Self {
        self.transitions.push(WorkflowTransition {
            from: from.into(),
            to: to.into(),
            trigger: trigger.into(),
        });
        self
    }

    /// Set the initial state
    pub fn set_initial_state(mut self, state: impl Into<String>) -> Self {
        self.initial_state = Some(state.into());
        self
    }

    /// Build the workflow definition (returns data that can be used with a client)
    pub fn build(self) -> WorkflowDefinition {
        WorkflowDefinition {
            name: self.name,
            description: self.description,
            states: self.states,
            transitions: self.transitions,
            initial_state: self.initial_state,
        }
    }
}

#[derive(Debug, Clone)]
pub struct WorkflowDefinition {
    pub name: String,
    pub description: Option<String>,
    pub states: Vec<WorkflowState>,
    pub transitions: Vec<WorkflowTransition>,
    pub initial_state: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowState {
    pub name: String,
    pub state_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowTransition {
    pub from: String,
    pub to: String,
    pub trigger: String,
}
